Improve type mapping documentation for Druid connector

tlblessing opened the pull request:
## Description

Improve type mapping documentation for Druid connector. Added a type mapping table based on Druid documentation.

> Is this change a fix, improvement, new feature, refactoring, or other?

Docs improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Connector type mapping documentation

> How would you describe this change to a non-technical end user or system administrator?

Added a data type mapping section


<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

jhlodin commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,73 @@ name from the properties file.
+The connector maps Druid types to the corresponding Trino types according to the
+following table:
+
+.. list-table:: Druid type to Trino type mapping
+  :widths: 30, 30, 50
+  :header-rows: 1
```
Several of these entries are one-to-many relationships which doesn't make sense in the context of type mapping, so I expect we'll be removing or changing a number of these.

hashhar replied:
Note that https://druid.apache.org/docs/latest/querying/sql-data-types.html talks about Druid's own SQL queries - not how Druid data types get exposed to other systems which is probably a source of common confusion.

tlblessing replied:
@hashhar should i remove that link?

jhlodin replied:
+1 to removing the sentence and link on line 100. If that document is confusing, and the newly-updated table is complete and correct, then I don't think it's necessary.

tlblessing replied:
Done

hashhar commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,73 @@ name from the properties file.
+  * - Druid type
+    - Trino type
+    - Notes
+  * - ``STRING``
```
IIRC we order these in a specific order. Please change to match that order.

Here's the mappings:

### Druid to Trino

`STRING` -> `VARCHAR`
`FLOAT` -> `REAL`
`DOUBLE` -> `DOUBLE`
`LONG` -> `BIGINT` (except the special `__time` column)
`TIMESTAMP` -> `TIMESTAMP` (ONLY the special `__time` column)

One thing which I think is worth calling out is that Druid doesn't have a real `NULL` value for any data type - by default it just treats `NULL` as the default value for a data type (e.g. LONG would be 0, DOUBLE would be 0.0, STRING would be `` (empty string) etc.).

For the reverse direction (Trino to Druid types) consider that Druid is a read-only connector so the scenarios where the write mappings get used are very very limited. It's the exact inverse of the read mappings however.

tlblessing replied:
@hashhar and @jhlodin since it is a read-only connector, I assume we don't include the reverse table? Please review again. Thank you.

jhlodin replied:
@hashhar We have the line "No other data types are supported", so just to clarify - are those five the only data types supported in Druid?

jhlodin replied:
@tlblessing I believe the reverse table is still needed. For those limited cases where the other direction is applicable, again per the "no other data types are supported" caveat, we need to be complete. Otherwise we're suggesting that Druid doesn't support most Trino data types.

hashhar replied:
Yes, only those 5 types are supported. And yes a reverse table is ok to add with the exact same mapping but in reverse.

tlblessing replied:
@hashhar and @jhlodin added the reverse mapping table. Ready for another review.

jhlodin replied:
>  Druid is a read-only connector so the scenarios where the write mappings get used are very very limited.

@hashhar Follow up question regarding what you said above in this thread. Since the Druid connector is read-only, what are these "very very limited" cases where a Trino to Druid type mapping applies? We want to be complete in the type mapping overview section that we're writing (https://github.com/trinodb/trino/pull/13738) to be sure we're accurate in describing where each mapping applies (i.e. not "read" and "write" mappings, because it's more accurately "read/write" and "DDL" mappings).

If there's no practical uses for a Trino-to-source mapping section in a read-only connector like Druid, @mosabua and I think we should leave it out. We just want to make sure that's accurate.

hashhar replied:
It will get used whenever Trino needs to "write" a query to be sent to Druid. i.e. WHERE clauses, JOIN conditions etc.

But for all practical purposes that mechanism is hidden from the user and there's nothing they can do to affect it so even if you explain the non-write cases where it gets used it doesn't actually help them other than maybe feel like they have some "knowledge".

cc: @ebyhr for opinion

ebyhr replied:
It's hidden and also I guess most users think it's part of "read". Documenting "Trino type to {remote} type mapping" in read-only connectors is little confusing in my opinion. Just removing the table looks good to me.

tlblessing replied:
Removed read table and adjusted boilerplate language to remove mention of the read direction.

mosabua replied:
Thanks for clarifying .. we will do that for all connectors as applicable.

Question still though .. I don't understand how the type mapping comes into play when "writing" a query for where clauses or joins ... it would only translate Trino SQL into the underlying SQL or whatever language of the data source .. how does type mapping affect that? Are different version created and sent depending on what the mapping is ? Can you think of an example?

hashhar replied:
e.g. With Druid a query like `WHERE __time = TIMESTAMP '2020-01-01 01:01:01'` uses the write mapping to convert the `TIMESTAMP` literal on Trino side to a value to be sent to Druid.

With a query like `SELECT __time FROM druid.default.table` the read mapping is used to convert values read from Druid to Trino values.

jhlodin commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -75,49 +75,26 @@ following table:
-      ``approxHistogram``, and so forth.
+    - Only applicable to the special ``_time`` column
+
+Note that Druid does not have a real ``NULL`` value for any data type. By
```
Remove "Note that"

tlblessing replied:
Done

jhlodin commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,50 @@ name from the properties file.
+For more information, see `Druid data types
+<https://druid.apache.org/docs/latest/querying/sql-data-types.html>`_.
+
+No other data types are supported.
```
Move this line to immediately below the table

tlblessing replied:
Done

hashhar commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,81 @@ name from the properties file.
+    -
+  * - ``BIGINT``
+    - ``LONG``
+    - Except for the special ``_time`` column
```
```suggestion
    - Except for the special ``_time`` column which is mapped as ``TIMESTAMP``
```

hashhar replied:
Or something like "see below"?

tlblessing replied:
incorporated with an edit (need punctuation)

hashhar commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,81 @@ name from the properties file.
+  :widths: 30, 30, 50
+  :header-rows: 1
+
+  * - Druid type
+    - Trino type
```
This looks reversed. `VARCHAR` is the Trino type, `STRING` is the Druid type.

tlblessing replied:
Done.

tlblessing commented:
@hashhar just one remaining question for you here https://github.com/trinodb/trino/pull/13776#discussion_r954234314.  Thank you  Ashhar, @ebyhr , and @jhlodin for your guidance and reviews. Incorporated comments, squashed commits , and ready for another review.

jhlodin commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,49 @@ name from the properties file.
 ------------
 
+Because Trino and Druid each support types that the other does not, this
+connector modifies some types when reading data. Refer to
+the following table for type mapping.
```
Remove "Refer to the following table for type mapping", it seems overkill when there's only one table (which is introduced below anyway)

tlblessing replied:
Done

hashhar commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,45 @@ name from the properties file.
+  * - ``STRING``
+    - ``VARCHAR``
+    -
+  * - ``REAL``
+    - ``FLOAT``
```
Druid has ``FLOAT`` mapped to Trino ``REAL``. So this should be reversed.

hashhar commented on docs/src/main/sphinx/connector/druid.rst:
```diff
@@ -57,6 +57,45 @@ name from the properties file.
+  * - ``DOUBLE``
+    - ``DOUBLE``
+    -
+  * - ``BIGINT``
+    - ``LONG``
```
Same here. Druid has ``LONG`` mapped to Trino ``BIGINT``. So this should be reversed.

hashhar reviewed (approved):
LGTM % 2 more swaps.

mosabua reviewed (approved):
Looks good to me now. Thank you.

tlblessing commented:
This PR is now ready to merge.
